mbf_costmap_nav/src/mbf_costmap_nav/free_pose_search.cpp

```diff
@@ -212,64 +223,71 @@ SearchSolution FreePoseSearch::findValidOrientation(const costmap_2d::Costmap2D&
 
 SearchSolution FreePoseSearch::search() const
 {
+  // restart markers
+  if (viz_)
+  {
+    viz_->deleteMarkers();
+  }
+
   const auto costmap2d = costmap_.getCostmap();
 
   // lock costmap so content doesn't change while adding cell costs
   boost::unique_lock<costmap_2d::Costmap2D::mutex_t> lock(*(costmap2d->getMutex()));
 
   std::unordered_set<int> in_queue_or_visited;
+  std::priority_queue<Cell, std::vector<Cell>, decltype(compare_strategy_)> queue(compare_strategy_);
 
   const std::vector<geometry_msgs::Point> footprint =
       safetyPadding(costmap_, config_.use_padded_fp, config_.safety_dist);
 
   SearchSolution sol;
+  Cell test_cell;
   sol.pose.theta = config_.goal.theta;
   bool outside_or_unknown{ false };
 
-  // initializing queue with the goal cell
-  Cell start;
-  if (costmap2d->worldToMap(config_.goal.x, config_.goal.y, start.x, start.y))
+  // enforce bounds if goal is outside the map
+  int test_cell_x, test_cell_y;
+  costmap2d->worldToMapEnforceBounds(config_.goal.x, config_.goal.y, test_cell_x, test_cell_y);
+  test_cell.x = static_cast<unsigned int>(test_cell_x);
+  test_cell.y = static_cast<unsigned int>(test_cell_y);
+  test_cell.cost = costmap2d->getCost(test_cell.x, test_cell.y);
+
+  // the cell center might be different from the goal pose (depends on costmap resolution)
+  geometry_msgs::Pose2D test_pose;
+  costmap2d->mapToWorld(test_cell.x, test_cell.y, test_pose.x, test_pose.y);
+
+  int cell_no_bounds_x, cell_no_bounds_y;
+  costmap2d->worldToMapNoBounds(config_.goal.x, config_.goal.y, cell_no_bounds_x, cell_no_bounds_y);
+  // if goal in bounds, we start the search from the goal pose and continue from the goal cell's center
+  if (cell_no_bounds_x == test_cell_x && cell_no_bounds_y == test_cell_y)
```

Comment:
```suggestion
  // if goal in bounds, we start the search from the goal pose and continue from the goal cell's center
  unsigned int dummy_x, dummy_y;
  if (!costmap2d->worldToMap(config_.goal.x, config_.goal.y, dummy_x, dummy_y))
```
S: Use the version that returns a bool when it's outside the bounds instead of comparing the cell coords (why does this version receive uint but enforce bounds receives int :joy:)